### PR TITLE
Deploy distributables to testnet guide

### PR DESCRIPTION
**Pull request by yhaspel**
In this commit:
* Publish App Installers to AWS bucket
* Generating a publish list text file with file / hash which is also published to AWS bucket

### CODE REVIEW

File: app/screens/wallet/TwelveWordsBackup.js [hunk 1]
```diff
@@ -119,7 +118,7 @@ class TwelveWordsBackup extends Component<Props, State> {
 
   twelveWordsPrint: string;
 
-  constructor(props) {
+  constructor(props: Props) {
```
Comment (IlyaVi) [inline app/screens/wallet/TwelveWordsBackup.js]: revert

File: app/components/contacts/AddNewContact.js [hunk 2]
```diff
@@ -95,7 +94,7 @@ class AddNewContact extends Component<Props, State> {
 
   state = { ...this.initialState };
 
-  static getDerivedStateFromProps(props, prevState) {
+  static getDerivedStateFromProps(props: Props, prevState: State) {
```
Comment (IlyaVi) [inline app/components/contacts/AddNewContact.js]: revert

File: app/components/contacts/AddNewContact.js [hunk 3]
```diff
@@ -119,7 +118,7 @@ class AddNewContact extends Component<Props, State> {
 
   renderFields = () => {
     const { defaultAddress } = this.props;
-    return fields.map(({ title, placeholder, fieldName, errorFieldName }) => {
+    return fields.map<Function>(({ title, placeholder, fieldName, errorFieldName }) => {
```
Comment (IlyaVi) [inline app/components/contacts/AddNewContact.js]: revert

File: .gitignore [hunk 4]
```diff
@@ -36,3 +37,5 @@ ct-libc/
 /desktop/main.prod.js
 /desktop/main.prod.js.map
 /node/
+/testnet-guide
```
Comment (IlyaVi) [inline .gitignore]: this folder exists?

File: .gitignore [hunk 5]
```diff
@@ -18,6 +18,7 @@
 .env.test.local
 .env.production.local
 .eslintcache
+spacemesh.json
```
Comment (IlyaVi) [inline .gitignore]: this file exists?

File: .travis.yml [hunk 6]
```diff
@@ -33,6 +33,8 @@ script:
   - npm run lint-ci
   - npm run build-dll
   - |
-    if [ "$TRAVIS_OS_NAME" == "osx" ]; then 
-      npm run package-all;
+    if [ $TRAVIS_OS_NAME == "osx" ]; then 
+      rm -f release/auth.md
```
Comment (IlyaVi) [inline .travis.yml]: why you remove something that doesn't exist?

File: .travis.yml [hunk 7]
```diff
@@ -33,6 +33,8 @@ script:
...
+      rm -f release/publishFilesList.txt
```
Comment (IlyaVi) [inline .travis.yml]: see previous comment

File: configs/packageConfig.js [hunk 8]
```diff
@@ -1,21 +1,89 @@
...
+  return hash;
+};
+
+const generateFile = async ({ destination, lines }) => {
+  try {
+    if (!fs.existsSync(destination)) {
+      await writeFileAsync(destination, lines.join('\n'));
+    }
+  } catch (error) {
+    throw error;
```
Comment (IlyaVi) [inline configs/packageConfig.js]: process.exit(1) instead of throw error

File: configs/packageConfig.js [hunk 9]
```diff
@@ -1,21 +1,89 @@
...
+  const shasum = crypto.createHash('sha512');
+  const fileContent = await readFileAsync(filename);
+  const hashsum = shasum.update(fileContent);
+  const hash = hashsum.digest('hex');
+  return hash;
+};
+
+const generateFile = async ({ destination, lines }) => {
+  try {
+    if (!fs.existsSync(destination)) {
```
Comment (IlyaVi) [inline configs/packageConfig.js]: generateFile is called once, thus file doesn't exist and this check is redundant

File: .gitignore (hunk 4 above)
Comment (yhaspel) [reply .gitignore]: Not anymore. Removing this.

File: .gitignore (hunk 5 above)
Comment (yhaspel) [reply .gitignore]: not anymore. was related to the google bucket upload. removing this.

File: app/components/contacts/AddNewContact.js (hunk 2 above)
Comment (yhaspel) [reply app/components/contacts/AddNewContact.js]: ok

File: app/screens/wallet/TwelveWordsBackup.js (hunk 1 above)
Comment (yhaspel) [reply app/screens/wallet/TwelveWordsBackup.js]: ok

File: app/components/contacts/AddNewContact.js (hunk 3 above)
Comment (yhaspel) [reply app/components/contacts/AddNewContact.js]: ok

File: .travis.yml (hunk 6 above)
Comment (yhaspel) [reply .travis.yml]: right.

File: .travis.yml (hunk 7 above)
Comment (yhaspel) [reply .travis.yml]: This file does exist once package-all has completed its run.

File: configs/packageConfig.js (hunk 8 above)
Comment (yhaspel) [reply configs/packageConfig.js]: ok

File: configs/packageConfig.js (hunk 9 above)
Comment (yhaspel) [reply configs/packageConfig.js]: ok

File: .travis.yml (hunk 7 above)
Comment (yhaspel) [reply .travis.yml]: ok

File: configs/packageConfig.js (hunk 9 above)
Comment (yhaspel) [reply configs/packageConfig.js]: generateFile will be called once per target batch.

File: configs/packageConfig.js [hunk 10]
```diff
@@ -1,21 +1,86 @@
+const fs = require('fs');
+const path = require('path');
+const util = require('util');
+const crypto = require('crypto');
+const readline = require('readline');
```
Comment (IlyaVi) [inline configs/packageConfig.js]: remove

Comment (yhaspel) [reply configs/packageConfig.js]: ok

File: configs/packageConfig.js [hunk 11]
```diff
@@ -66,9 +120,27 @@ const getBuildOptions = (target) => ({
...
+      bucket: 'app-binaries.spacemesh.io'
+    },
     directories: {
       buildResources: 'resources',
       output: 'release'
+    },
+    afterAllArtifactBuild: async (buildResult) => {
+      try {
+        const artifactsToPublishFile = path.join(__dirname, '..', 'release', 'publishFilesList.txt');
+        if (!fs.existsSync(artifactsToPublishFile)) {
```
Comment (IlyaVi) [inline configs/packageConfig.js]: move this outside the callback and put right after declaring "fileHashList" var

File: configs/packageConfig.js [hunk 12]
```diff
@@ -1,21 +1,75 @@
...
+      }
+    }
+  }
+  return lines;
+};
+
 const getBuildOptions = (target) => ({
   targets: Platform[target.toUpperCase()].createTarget(),
-  publish: 'never', 
+  publish: 'always',
```
Comment (IlyaVi) [inline configs/packageConfig.js]: change this to "onTag" option and test it

File: configs/packageConfig.js [hunk 13]
```diff
@@ -1,21 +1,75 @@
...
 
+const getFileHash = async ({ filename }) => {
+  const shasum = crypto.createHash('sha512');
+  const fileContent = await readFileAsync(filename);
+  const hashsum = shasum.update(fileContent);
+  const hash = hashsum.digest('hex');
+  return hash;
+};
+
+const getCompiledLines = async ({ artifactsToPublishFile, artifactPaths }) => {
```
Comment (IlyaVi) [inline configs/packageConfig.js]: rename

File: configs/packageConfig.js [hunk 14]
```diff
@@ -1,21 +1,75 @@
...
+  const acceptedSuffixes = ['dmg', 'exe', 'deb', 'snap', 'AppImage'];
+  const getArtifactNameAndSuffix = ({ fullPath }) => {
```
Comment (IlyaVi) [inline configs/packageConfig.js]: I really don't understand your obsession to declare inner functions, especially when outer function is called in loop

File: configs/packageConfig.js [hunk 15]
```diff
@@ -1,21 +1,75 @@
...
+    const artifactSuffixSplit = fullPath.split('.');
+    const artifactSuffix = artifactSuffixSplit[artifactSuffixSplit.length - 1];
+    const artifactNameSplit = fullPath.split('/');
+    const artifactName = artifactNameSplit[artifactNameSplit.length - 1];
+    return { artifactName, artifactSuffix };
+  };
+  const fileContent = await readFileAsync(artifactsToPublishFile, 'utf8');
+  const lines = fileContent.split('\n');
```
Comment (IlyaVi) [inline configs/packageConfig.js]: why splitting?? then you search for index of specific line, makes existence check and only then make replace. What the difference there is between this and just replacing in the original string? (Except the overcomplicating of course)

File: package.json [hunk 16]
```diff
@@ -123,6 +123,7 @@
     "electron-store": "^4.0.0",
     "grpc": "^1.21.1",
     "js-xdr": "^1.1.2",
+    "lodash": ">=4.17.14",
```
Comment (IlyaVi) [inline package.json]: this is dev dependency

File: configs/packageConfig.js [hunk 17]
```diff
@@ -1,21 +1,75 @@
...
+  for (const fullPath of artifactPaths) {
+    const { artifactName, artifactSuffix } = getArtifactNameAndSuffix({ fullPath });
+    // installers only
+    if (acceptedSuffixes.indexOf(artifactSuffix) >= 0) {
+      const installerKey = `${artifactSuffix}_installer`;
+      const shaKey = `${artifactSuffix}_sha512`;
+      const lineIndex = lines.findIndex((editedLine) => editedLine.includes(installerKey));
+      if (lineIndex >= 0) {
+        const hash = await getFileHash({ filename: fullPath });
+        const installerRegex = new RegExp(installerKey, 'g');
```
Comment (IlyaVi) [inline configs/packageConfig.js]: also you don't need to create regex, replace works fine in this case

File: configs/packageConfig.js (hunk 11 above)
Comment (yhaspel) [reply configs/packageConfig.js]: ok

File: configs/packageConfig.js (hunk 12 above)
Comment (yhaspel) [reply configs/packageConfig.js]: ok

File: configs/packageConfig.js (hunk 13 above)
Comment (yhaspel) [reply configs/packageConfig.js]: ok

File: configs/packageConfig.js (hunk 14 above)
Comment (yhaspel) [reply configs/packageConfig.js]: OK. I've moved the functionality to the loop. Not sure why this is an issue in your view.

File: configs/packageConfig.js (hunk 17 above)
Comment (yhaspel) [reply configs/packageConfig.js]: ok

File: configs/packageConfig.js (hunk 15 above)
Comment (yhaspel) [reply configs/packageConfig.js]: Changed to json

File: app/App.js [hunk 18]
```diff
@@ -1,3 +1,5 @@
+// @flow
+import { hot, setConfig } from 'react-hot-loader';
```
Comment (IlyaVi) [inline app/App.js]: move hot reload to index.js where it was initially

File: app/screens/wallet/Backup.js [hunk 19]
```diff
@@ -1,4 +1,3 @@
-// @flow
```
Comment (IlyaVi) [inline app/screens/wallet/Backup.js]: return flow back

File: app/screens/wallet/Overview.js [hunk 20]
```diff
@@ -1,4 +1,3 @@
-// @flow
```
Comment (IlyaVi) [inline app/screens/wallet/Overview.js]: same as above

File: app/screens/wallet/SendCoins.js [hunk 21]
```diff
@@ -1,4 +1,3 @@
-// @flow
```
Comment (IlyaVi) [inline app/screens/wallet/SendCoins.js]: same as above

File: app/screens/wallet/TestMe.js [hunk 22]
```diff
@@ -1,4 +1,3 @@
-// @flow
```
Comment (IlyaVi) [inline app/screens/wallet/TestMe.js]: same as above

File: app/screens/wallet/TwelveWordsBackup.js [hunk 23]
```diff
@@ -1,4 +1,3 @@
-// @flow
```
Comment (IlyaVi) [inline app/screens/wallet/TwelveWordsBackup.js]: same as above

File: configs/packageConfig.js [hunk 24]
```diff
@@ -1,21 +1,100 @@
+// @flow
```
Comment (IlyaVi) [inline configs/packageConfig.js]: remove flow

File: configs/packageConfig.js [hunk 25]
```diff
@@ -66,9 +145,25 @@ const getBuildOptions = (target) => ({
...
+      provider: 's3',
+      bucket: 'app-binaries.spacemesh.io'
+    },
     directories: {
       buildResources: 'resources',
       output: 'release'
+    },
+    afterAllArtifactBuild: async (buildResult) => {
+      try {
+        const artifactsToPublishFile = path.join(__dirname, '..', 'release', 'publishFilesList.json');
```
Comment (IlyaVi) [inline configs/packageConfig.js]: move this and next line out of callback to create file once

File: configs/packageConfig.js [hunk 26]
```diff
@@ -66,9 +145,25 @@ const getBuildOptions = (target) => ({
...
+        await generateFileHashFile({ destination: artifactsToPublishFile });
+        const compliledHashList = await getCompiledHashList({ artifactsToPublishFile, artifactPaths: buildResult.artifactPaths });
+        await writeFileAsync(artifactsToPublishFile, JSON.stringify(compliledHashList));
```
Comment (IlyaVi) [inline configs/packageConfig.js]: move write to getCompiledHashList as you already read file in this function

File: configs/packageConfig.js [hunk 27]
```diff
@@ -1,21 +1,100 @@
...
+  return hash;
+};
+
+const getCompiledHashList = async ({ artifactsToPublishFile, artifactPaths }) => {
+  const acceptedSuffixes = ['dmg', 'exe', 'deb', 'snap', 'AppImage'];
+  const fileContent = await readFileAsync(artifactsToPublishFile);
+  const hashList = JSON.parse(fileContent);
+  for (const fullPath of artifactPaths) {
+    const artifactSuffixSplit = fullPath.split('.');
+    const artifactSuffix = artifactSuffixSplit[artifactSuffixSplit.length - 1];
```
Comment (IlyaVi) [inline configs/packageConfig.js]: artifactSuffixSplit has no usage further so simplify those 2 lines 
const artifactSuffix = fullPath.split('.').pop()

File: configs/packageConfig.js [hunk 28]
```diff
@@ -1,21 +1,100 @@
...
+    const artifactNameSplit = fullPath.split('/');
+    const artifactName = artifactNameSplit[artifactNameSplit.length - 1];
```
Comment (IlyaVi) [inline configs/packageConfig.js]: same as above

File: app/App.js (hunk 18 above)
Comment (yhaspel) [reply app/App.js]: ok

File: app/screens/wallet/Backup.js (hunk 19 above)
Comment (yhaspel) [reply app/screens/wallet/Backup.js]: ok

File: configs/packageConfig.js [hunk 29]
```diff
@@ -1,21 +1,100 @@
...
+    if (acceptedSuffixes.indexOf(artifactSuffix) >= 0) {
+      const installerKey = `${artifactSuffix}_installer`;
+      const shaKey = `${artifactSuffix}_sha512`;
+      let targetKey = null;
+      for (const key in hashList) {
+        if (hashList[key].file.includes(installerKey)) {
+          targetKey = key;
+        }
+      }
+      if (!!targetKey) {
```
Comment (IlyaVi) [inline configs/packageConfig.js]: targetKey is not available in this scope

File: app/index.js [hunk 30]
```diff
@@ -1,26 +1,19 @@
...
-renderWrapper(App);
-
-if (module.hot) {
-  module.hot.accept('./App', () => {
-    // eslint-disable-next-line global-require
-    const NextApp = require('./App').default;
-    renderWrapper(NextApp);
-  });
-}
+renderWrapper(process.env.NODE_ENV === 'development' ? hot(module)(App) : App);
```
Comment (IlyaVi) [inline app/index.js]: simplify this, there is no need in wrapping function

File: configs/packageConfig.js [hunk 31]
```diff
@@ -1,21 +1,99 @@
...
+  const fileContent = await readFileAsync(filename);
+  const hashsum = shasum.update(fileContent);
+  const hash = hashsum.digest('hex');
+  return hash;
+};
+
+const compileHashListFile = async ({ artifactsToPublishFile, artifactPaths }) => {
+  const acceptedSuffixes = ['dmg', 'exe', 'deb', 'snap', 'AppImage'];
+  const fileContent = await readFileAsync(artifactsToPublishFile);
+  let hashList = JSON.stringify(JSON.parse(fileContent));
```
Comment (IlyaVi) [inline configs/packageConfig.js]: [hint](https://lmgtfy.com/?q=nodejs+read+file+to+string)

File: app/index.js (hunk 30 above)
Comment (yhaspel) [reply app/index.js]: ok

File: configs/packageConfig.js (hunk 31 above)
Comment (yhaspel) [reply configs/packageConfig.js]: That's just disrespectful.

**Pull request merged**